File: csharp/src/refactoring/parallel-change/sprout-change/SproutChange.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace RefactoringAvanzado.Refactoring.ParallelChange.SproutChange;

public class CartItem
{
    public string Id { get; init; } = string.Empty;
    public decimal Price { get; init; }
    public int Qty { get; init; }
    public string? Category { get; init; }
}

public enum Region
{
    US,
    EU
}

public interface ITaxPolicy
{

}

public static class TaxCalculator
{
    // Regla existente: un único impuesto plano por región; los libros y la comida están exentos en la UE
    public static decimal CalculateTotal(List<CartItem> cart, Region region)
    {
        var subtotal = cart.Sum(it => it.Price * it.Qty);

        var tax = CalculateTax(subtotal);

        return RoundCurrency(subtotal + tax);
    }

    public static decimal RoundCurrency(decimal amount)
    {
        return Math.Round(amount, 2);
    }

    public static decimal CalculateTax(decimal subtotal)
    {
        if (region == Region.US || region == Region.EU) {
             return NewCalculateTax(subtotal);
        }

         return OldCalculateTax(subtotal);
    }

    public static decimal NewCalculateTax(decimal subtotal)
    {
        decimal tax = 0;
        if (region == Region.US)
        {
            tax = subtotal * 0.17m;
        }
    }

    public static decimal OldCalculateTax(decimal subtotal)
    {
        decimal tax = 0;
        if (region == Region.US)
        {
            tax = subtotal * 0.07m; // 7% plano
        }
        else if (region == Region.EU)
        {
            // exenciones ingenuas en línea
            var taxable = cart
                .Where(it => it.Category != "books" && it.Category != "food")
                .Sum(it => it.Price * it.Qty);
            tax = taxable * 0.2m; // 20% plano solo sobre los ítems gravables
        }
        return tax;
   }
}

// Uso de ejemplo, mantenido simple para estudiantes
public static class SproutChangeDemo
{
    public static decimal DemoSprout()
    {
        var cart = new List<CartItem>
        {
            new() { Id = "p1", Price = 10, Qty = 2, Category = "general" },
            new() { Id = "b1", Price = 20, Qty = 1, Category = "books" }
        };
        return TaxCalculator.CalculateTotal(cart, Region.EU);
    }
}

File: typescript/src/refactoring/parallel-change/sprout-change/sprout-change.ts
export type CartItem = {
  id: string
  price: number
  qty: number
  category?: 'general' | 'books' | 'food'
}
export type Region = 'US' | 'EU'

interface TaxPolicy {

}
// Regla existente: un único impuesto plano por región; los libros y la comida están exentos en la UE
function calculateUSTax(cart: CartItem[]) {
  const usSubtotal = cart.reduce((s, it) => s + it.price * it.qty, 0)

  return usSubtotal * 0.07 // 7% plano
}

function calculateEUTax(cart: CartItem[]) {
  const taxable = cart
    .filter((it) => it.category !== 'books' && it.category !== 'food')
    .reduce((s, it) => s + it.price * it.qty, 0)
  return taxable * 0.2 // 20% plano solo sobre los ítems gravables
}

function calculateDefault(cart: CartItem[]) {
  return 0
}

// (reglas embebidas en línea)
export function calculateTotal(cart: CartItem[], region: Region): number {
  const subtotal = cart.reduce((s, it) => s + it.price * it.qty, 0)

  let tax = 0
  if (region === 'US') {
    tax = calculateUSTax(cart);
  } else if (region === 'EU') {
    // exenciones ingenuas en línea
    tax = calculateEUTax(cart);
  } else {
    tax = calculateDefault(cart);
  }

  return roundCurrency(subtotal + tax)
}

export function roundCurrency(amount: number): number {
  return Math.round(amount * 100) / 100
}

// Uso de ejemplo, mantenido simple para estudiantes
export function demoSprout(): number {
  const cart: CartItem[] = [
    {id: 'p1', price: 10, qty: 2, category: 'general'},
    {id: 'b1', price: 20, qty: 1, category: 'books'},
  ]
  return calculateTotal(cart, 'EU')
}
